controller/scheduler-v2/scheduler.go

```diff
@@ -69,62 +80,91 @@ func (s *Scheduler) Run() error {
 	log.Info("starting scheduler loop")
 	defer log.Info("exiting scheduler loop")
 
-	jobTicker := time.Tick(30 * time.Second)
-	formationTicker := time.Tick(time.Minute)
-	formationTimeout := time.After(time.Second)
-	go func() {
-		for {
-			select {
-			case <-jobTicker:
-				triggerChan(s.jobSync)
-			case <-formationTicker:
-				triggerChan(s.formationSync)
-			case <-formationTimeout:
-				triggerChan(s.formationSync)
-			}
-		}
-	}()
-	triggerChan(s.jobSync)
+	tickChannel(s.syncJobs, 30*time.Second)
+	tickChannel(s.syncFormations, time.Minute)
+	tickChannel(s.syncHosts, 5*time.Minute)
+
+	stream, err := s.StreamHosts(s.hostChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream hosts. Error: %v", err)
+	}
+	defer stream.Close()
+
+	stream, err = s.StreamFormations(nil, s.formationChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream formations. Error: %v", err)
+	}
+	defer stream.Close()
+
+	go s.RunPutJobs()
 
 	for {
+		// First handle events reconciling our state with the cluster
 		select {
 		case <-s.stop:
 			return nil
-		case req := <-s.jobRequests:
-			s.jobRequestHandler(req)
-		case <-s.rectifyJobs:
-			s.RectifyJobs()
+		case h := <-s.hostChange:
+			s.followHost(h)
+			continue
+		case he := <-s.hostEvents:
+			s.handleHostEvent(he)
+			continue
 		case ef := <-s.formationChange:
-			s.formationChangeHandler(ef)
-		case <-s.formationSync:
+			s.FormationChange(ef)
+			continue
+		case <-time.After(50 * time.Millisecond):
+		}
+
+		// Then handle entropy minimizing sync events
+		select {
+		case <-s.syncHosts:
+			s.SyncHosts()
+		case <-s.syncFormations:
 			s.SyncFormations()
-		case <-s.jobSync:
+		case <-s.syncJobs:
 			s.SyncJobs()
+		default:
+		}
+
+		// Then handle events that could mutate the cluster
+		if s.isLeader {
+			select {
+			case req := <-s.jobRequests:
+				s.HandleJobRequest(req)
+			case <-s.rectifyJobs:
+				s.RectifyJobs()
+			default:
+			}
 		}
 	}
 	return nil
 }
 
-func (s *Scheduler) SyncJobs() (err error) {
-	fLog := s.log.New("fn", "Sync")
+func (s *Scheduler) SyncHosts() (err error) {
+	log := s.log.New("fn", "SyncHosts")
 
-	defer func() {
-		s.sendEvent(NewEvent(EventTypeClusterSync, err, nil))
-	}()
+	defer s.sendEvent(NewEvent(EventTypeHostSync, err, nil))
```

Comment:
This needs to be enclosed in an anonymous function so that `err` is evaluated when the defer is called (otherwise it will always be `nil`).